includes/Core/Authentication/Token.php

```diff
@@ -0,0 +1,147 @@
+<?php
+/**
+ * Class Google\Site_Kit\Core\Authentication\Token
+ *
+ * @package   Google\Site_Kit
+ * @copyright 2021 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\Authentication;
+
+use Google\Site_Kit\Core\Storage\User_Options_Interface;
+use Google\Site_Kit\Core\Storage\Encrypted_User_Options;
+use Google\Site_Kit\Core\Authentication\Clients\OAuth_Client;
+
+/**
+ * Class representing the OAuth token for a user.
+ *
+ * This includes the access token, its creation and expiration data, and the refresh token.
+ * This class is compatible with `Google\Site_Kit\Core\Storage\User_Setting`, as it should in the future be adjusted
+ * so that the four pieces of data become a single user setting.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+final class Token {
+
+	/**
+	 * User_Options_Interface implementation.
+	 *
+	 * @since n.e.x.t
+	 * @var User_Options_Interface
+	 */
+	protected $user_options;
+
+	/**
+	 * Encrypted_User_Options instance.
+	 *
+	 * @since 1.0.0
+	 * @var Encrypted_User_Options
+	 */
+	private $encrypted_user_options;
+
+	/**
+	 * User_Setting constructor.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param User_Options_Interface $user_options User_Options_Interface instance.
+	 */
+	public function __construct( User_Options_Interface $user_options ) {
+		$this->user_options           = $user_options;
+		$this->encrypted_user_options = new Encrypted_User_Options( $this->user_options );
+	}
+
+	/**
+	 * Checks whether or not the setting exists.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return bool True on success, false on failure.
+	 */
+	public function has() {
+		if ( ! $this->get() ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Gets the value of the setting.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return mixed Value set for the option, or default if not set.
+	 */
+	public function get() {
+		$access_token = $this->encrypted_user_options->get( OAuth_Client::OPTION_ACCESS_TOKEN );
+		if ( empty( $access_token ) ) {
+			return array();
+		}
+
+		$token = array(
+			'access_token' => $access_token,
+			'expires_in'   => (int) $this->user_options->get( OAuth_Client::OPTION_ACCESS_TOKEN_EXPIRES_IN ),
+			'created'      => (int) $this->user_options->get( OAuth_Client::OPTION_ACCESS_TOKEN_CREATED ),
+		);
+
+		$refresh_token = $this->encrypted_user_options->get( OAuth_Client::OPTION_REFRESH_TOKEN );
+		if ( ! empty( $refresh_token ) ) {
+			$token['refresh_token'] = $refresh_token;
+		}
```

Comment:
This implementation is in line with how the OAuth client works. The token data only includes a `refresh_token` key if there is one, so I think it makes sense to mirror that. Also, without an `access_token` the entire data is invalid, so I don't think we should in that case return that state.

Reply:
Ah, you're referring to `Google\Client` correct? If so, I think I see what you mean.

We could use `null` for the empty value here which would be interpreted the same within the `Client` because it uses `isset` everywhere.

If not, then this is fine 👍

Reply:
I think it makes sense to keep as is to follow the client behavior - technically `null` would work, but I guess with `array_key_exists` being a thing it is still technically a slightly different format.